tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/TraceAnnotations/TraceAnnotationsIntegration.cs

```diff
@@ -0,0 +1,114 @@
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <param name="methodHandle">The RuntimeMethodHandle representing the instrumented method</param>
+        /// <param name="typeHandle">The RuntimeTypeHandle representing the instrumented method's owning type</param>
+        /// <returns>Calltarget state value</returns>
+        internal static CallTargetState OnMethodBegin<TTarget>(TTarget instance, RuntimeMethodHandle methodHandle, RuntimeTypeHandle typeHandle)
+        {
+            var info = InstrumentedMethodCache.GetOrAdd(
+                    new RuntimeHandleTuple(methodHandle, typeHandle),
+                    key => TraceAnnotationInfoFactory.Create(MethodBase.GetMethodFromHandle(key.MethodHandle, key.TypeHandle)));
+
+            var tags = new TraceAnnotationTags();
+            var scope = Tracer.Instance.StartActiveInternal(info.OperationName, tags: tags);
+            scope.Span.ResourceName = info.ResourceName;
+
+            Tracer.Instance.TracerManager.Telemetry.IntegrationGeneratedSpan(IntegrationId);
+
+            return new CallTargetState(scope);
+        }
+
+        /// <summary>
+        /// OnMethodEnd callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <param name="exception">Exception instance in case the original code threw an exception.</param>
+        /// <param name="state">Calltarget state value</param>
+        /// <returns>A default CallTargetReturn to satisfy the CallTarget contract</returns>
+        internal static CallTargetReturn OnMethodEnd<TTarget>(TTarget instance, Exception exception, in CallTargetState state)
+        {
+            state.Scope.DisposeWithException(exception);
+            return CallTargetReturn.GetDefault();
+        }
+
+        /// <summary>
+        /// OnMethodEnd callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <typeparam name="TReturn">Type of the return value</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <param name="returnValue">Return value</param>
+        /// <param name="exception">Exception instance in case the original code threw an exception.</param>
+        /// <param name="state">Calltarget state value</param>
+        /// <returns>A response value, in an async scenario will be T of Task of T</returns>
+        internal static CallTargetReturn<TReturn> OnMethodEnd<TTarget, TReturn>(TTarget instance, TReturn returnValue, Exception exception, in CallTargetState state)
+        {
+            // If the return type is not Task or Task`1 then we can dispose the scope now, otherwise the scope will be disposed when OnAsyncMethodEnd is invoked
+            if (returnValue is not Task)
```

Comment:
Nice, and `ValueTasks`?

Reply:
🤦🏼 Forgot about those! Done in 158ca48ce2724f4fcde9e150182e4c4750f8fce8

`ValueTask` is available as early as [netcoreapp2.1](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask?view=netcore-2.1) and `ValueTask<TResult>` is available as early as [netcoreapp1.0](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask-1?view=netcore-1.0), so we'll need additional work to hook up the OnAsyncMethod callback for our earlier build targets. I've logged a JIRA ticket to track this